tests/nn/beam_search_test.py

```diff
@@ -12,120 +12,98 @@
+    ]
 )
 
 # A transition matrix that favors repeated ngrams
-repeated_ngram_transition_probabilities = torch.tensor(
-    [
-        [0.0, 1.0, 0.0, 0.0, 0.0, 0.0],  # start token -> jth token
-        [0.0, 0.0, 0.4, 0.6, 0.0, 1e-9],  # 1st token -> jth token
-        [0.0, 0.0, 0.0, 1.0, 0.0, 1e-9],  # 2nd token -> jth token
-        [0.0, 1.0, 0.0, 0.0, 0.0, 1e-9],  # ...
-        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],  # not used
-        [0.0, 0.0, 0.0, 0.0, 0.0, 1.0],
-    ]  # end token -> jth token
+repeated_ngram_transition_probabilities_0 = torch.tensor(
+    [  # START 1    2    3    4   END
+        [0.0, 1.0, 0.0, 0.0, 0.0, 0.0],   # START -> j
+        [0.0, 0.0, 0.4, 0.6, 0.0, 1e-9],  # 1 -> j
+        [0.0, 0.0, 0.0, 1.0, 0.0, 1e-9],  # 2 -> j
+        [0.0, 1.0, 0.0, 0.0, 0.0, 1e-9],  # 3 -> j
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],   # 4 -> j (not used)
+        [0.0, 0.0, 0.0, 0.0, 0.0, 1.0],   # END -> j (doesn't matter)
+    ]
 )
 
+# Another transition matrix that favors repeated ngrams
+repeated_ngram_transition_probabilities_1 = torch.tensor(
+    [  # START 1    2    3    4   END
+        [0.0, 0.4, 0.3, 0.2, 0.1, 0.0],  # START -> j
+        [0.0, 0.4, 0.3, 0.2, 0.1, 0.1],  # 1 -> j
+        [0.0, 0.0, 0.4, 0.3, 0.2, 0.1],  # 2 -> j
+        [0.0, 0.0, 0.3, 0.4, 0.2, 0.1],  # 3 -> j
+        [0.0, 0.0, 0.2, 0.3, 0.4, 0.1],  # 4 -> j
+        [0.2, 0.1, 0.2, 0.2, 0.2, 0.1],  # END -> j (doesn't matter)
+    ]
+)
+# fmt: on
+
 log_probabilities = torch.log(
     torch.tensor([[0.1, 0.3, 0.3, 0.3, 0.0, 0.0], [0.0, 0.0, 0.4, 0.3, 0.2, 0.1]])
 )
 
 
-def take_step_no_timestep(
-    last_predictions: torch.Tensor, state: Dict[str, torch.Tensor]
-) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
-    """
-    Take decoding step.
-
-    This is a simple function that defines how probabilities are computed for the
-    next time step during the beam search.
-
-    We use a simple target vocabulary of size 6. In this vocabulary, index 0 represents
-    the start token, and index 5 represents the end token. The transition probability
-    from a state where the last predicted token was token `j` to new token `i` is
-    given by the `(i, j)` element of the matrix `transition_probabilities`.
-    """
-    log_probs_list = []
-    for last_token in last_predictions:
-        log_probs = torch.log(transition_probabilities[last_token.item()])
-        log_probs_list.append(log_probs)
-
-    return torch.stack(log_probs_list), state
-
+def get_step_function(
+    transition_matrix: torch.Tensor, with_timestep: bool = False
+) -> Union[StepFunctionTypeNoTimestep, StepFunctionTypeWithTimestep]:
+    def _step_function(
+        last_predictions: torch.Tensor, state: Dict[str, torch.Tensor]
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        log_probs_list = []
+        for last_token in last_predictions:
+            log_probs = torch.log(transition_matrix[last_token.item()])
+            log_probs_list.append(log_probs)
```

Comment:
It seems to me that there should be a way to do this without a for loop? Not that it really matters, but something like `Tensor.index_select(...)`?